File: 0114-flatten-binary-tree-to-linked-list/solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def flatten(self, root):
        """
        :type root: TreeNode
        :rtype: None Do not return anything, modify root in-place instead.
        """
        def isLeaf(node):
            return (node and (not node.left) and (not node.right))
        
        if not root:
            return root
        
        if root.left and isLeaf(root.left):
            if isLeaf(root.right):
                temp = root.right
                root.right = root.left
                root.right.right = temp
                root.left = None
            
            elif not root.right:
                root.right = root.left 
                root.left = None
            
        elif not root.left:
            self.flatten(root.right)
            return root
                
            
        firstHalf = self.flatten(root.left)
        secondHalf = self.flatten(root.right)
        if firstHalf != None:
            root.right = firstHalf
            cur = firstHalf
            while cur and cur.right:
                cur = cur.right

            cur.right = secondHalf
            root.left = None
        else:
            root.right = secondHalf
        
        return root

File: 0114-flatten-binary-tree-to-linked-list/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isLeaf(TreeNode* node)
    {
        return (node && node->left == nullptr && node->right == nullptr);
    }
    void flatten(TreeNode* root) 
    {
        if(root == nullptr) return;   
        if(root->left == nullptr && root->right == nullptr) return;
        else if(root->left == nullptr && isLeaf(root->right)) return;
        
        // cout<<root->val<<endl;
        TreeNode* leftNode  = root->left;
        TreeNode* rightNode = root->right;
        
        root->left = nullptr;
        root->right = nullptr;
        
        flatten(leftNode);
        flatten(rightNode);
        
        TreeNode* cur = leftNode;
        
        
        root->right = leftNode;
        if(root->right)
        {
            while(cur != nullptr && cur->right != nullptr)
            {
                cur = cur->right;
            }
            cur->right  = rightNode;
        }
        else
        {
            root->right = rightNode;
        }
        
        
        return;
        
    }
};
